src/dmd/dcache.d

```diff
@@ -0,0 +1,376 @@
+/**
+ * Compiler implementation of the
+ * $(LINK2 http://www.dlang.org, D programming language).
+ *
+ * Copyright:   Copyright (c) 1999-2017 by Digital Mars, All Rights Reserved
+ * License:     $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Source:      $(LINK2 https://github.com/dlang/dmd/blob/master/src/dmd/dcache.d, _dcache.d)
+ *
+ * Dcache - an implementation if a distributed cache over memory-mapped file.
+ * It doesn't try to be smart and has most basic interface in the way of get/put,
+ * both operations are synchronious and "atomic" by using a simple spin-lock.
+ * 
+ * Key is assumed to be a strong hash (e.g. MD5) of some object and its first bits are used directly.
+ *
+ * Intended use
+ *
+ * Allow DMD instances (and frontends on its base) to share CTFE computation results and such,
+ * importantly a key must reflect the state of all dependent entities such as contents of files.
+ *
+ * TODOs:
+ * - eviction policy (e.g. time based expiration)
+ * - sizing policy (is 128 Mb enough for everybody?)
+ * - better hash map algorithm (it's really dumb)
+ * - better block management (does a linear pass to clean up entries ATM)
+ */
+module dmd.dcache;
+
+import core.sys.posix.sys.mman;
+import core.sys.posix.fcntl;
+import core.sys.posix.semaphore;
+import core.sys.posix.time;
+import core.sys.posix.unistd;
+import core.stdc.errno;
+import core.stdc.stdio;
+import core.stdc.stdlib;
+import core.stdc.string;
+import core.atomic;
+
+version(Posix)
+{
+    struct DCache
+    {
+        /// Load the cache identified by 'keyspace'
+        void initialize(const(char)[] keyspace)
+        {
+            char[512] buf;
+            const(char)* home = getenv("HOME");
```

Comment:
I have no idea what all the following string manipulation and file calls are doing without very carefully reading the code. Please add a block comment.